### PR TITLE
Improved MessageToMessageEncoder

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
MessageToMessageEncoder introduces an additional layer of queuing which duplicates the queueing at ChannelOutboundBuffer. Encoders which inherit from MessageToMessageEncoder could instead directly write to the channel instead of relying on MessageToMessageEncoder to do this for them.

Modifications:
- Introduce a light weight promise aggregator which can wrap the original promise from the ChannelOutboundHandler.write(..) method with the result of multiple Channel.write(..) operations done during the encode.
- Introduce a base class which does the type matching that MessageToMessageEncoder can inherit from to take advantage of the improved promise management.
- Update codec-redis to demonstrate the use case for this new encoder
- Update codec-http2 to use the new light weight promise aggregator.

Result:
Classes which currently inherit from MessageToMessageEncoder have an alternative class to inherit from which removes a layer of queueing.
MessageToMessageEncoder promise management is improved.
Fixes https://github.com/netty/netty/issues/5065

### CODE REVIEW

**Comment by Scottmitch** [comment]
@trustin @normanmaurer - PTAL

@jongyeol - You may also be interested in the redis encoder portion.

File: transport/src/main/java/io/netty/channel/DefaultChannelPromiseAggregatorFactory.java [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    private boolean doneAllocating;
+
+    /**
+     * Create a new instance which returns an appropriate type if {@code aggregationResult} is {@link #isVoid()}.
+     * @param aggregationResult The promise whose status will reflect the result of the aggregation.
+     * @param executor the {@link EventExecutor} which is used to notify the promise once it is complete.
+     */
+    public static ChannelPromiseAggregatorFactory newInstance(ChannelPromise aggregationResult,
+                                                              EventExecutor executor) {
+        return aggregationResult.isVoid() ? new VoidChannelPromiseAggregatorFactory(aggregationResult.channel(), true)
```
Comment (netkins) [inline transport/src/main/java/io/netty/channel/DefaultChannelPromiseAggregatorFactory.java]: MAJOR Ternary operators that can be simplified with || or && [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3ASimplifiedTernary)

File: transport/src/main/java/io/netty/channel/DefaultChannelPromiseAggregatorFactory.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        ++expectedCount;
+        return this;
+    }
+
+    /**
+     * Signify that no more {@link #newPromise()} allocations will be made.
+     * The aggregation can not be successful until this method is called.
+     * @return The promise that is the aggregation of all promises allocated with {@link #newPromise()}.
+     */
+    public ChannelPromise doneAllocatingPromises() {
```
Comment (netkins) [inline transport/src/main/java/io/netty/channel/DefaultChannelPromiseAggregatorFactory.java]: MAJOR Add the "@Override" annotation above this method signature [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1161)

File: transport/src/main/java/io/netty/channel/DefaultChannelPromiseAggregatorFactory.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        assert !aggregationResult.isDone();
+        this.promise = aggregationResult;
+    }
+
+    /**
+     * Allocate a new promise which will be used to aggregate the overall success of this promise aggregator.
+     * @return A new promise which will be aggregated.
+     * {@code null} if {@link #doneAllocatingPromises()} was previously called.
+     */
+    public ChannelPromise newPromise() {
```
Comment (netkins) [inline transport/src/main/java/io/netty/channel/DefaultChannelPromiseAggregatorFactory.java]: MAJOR Add the "@Override" annotation above this method signature [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1161)

File: transport/src/main/java/io/netty/channel/DefaultChannelPromiseAggregatorFactory.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    }
+
+    @Override
+    public boolean trySuccess(Void result) {
+        if (awaitingPromises()) {
+            ++doneCount;
+            if (allPromisesDone()) {
+                return tryPromise();
+            }
+            // TODO: We break the interface a bit here, but we can't return false because according to the javadoc
```
Comment (netkins) [inline transport/src/main/java/io/netty/channel/DefaultChannelPromiseAggregatorFactory.java]: INFO Complete the task associated to this TODO comment. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)

File: transport/src/main/java/io/netty/channel/DefaultChannelPromiseAggregatorFactory.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+    @Override
+    public boolean tryFailure(Throwable cause) {
+        if (allowFailure()) {
+            ++doneCount;
+            lastFailure = cause;
+            if (allPromisesDone()) {
+                return tryPromise();
+            }
+            // TODO: We break the interface a bit here, but we can't return false because according to the javadoc
```
Comment (netkins) [inline transport/src/main/java/io/netty/channel/DefaultChannelPromiseAggregatorFactory.java]: INFO Complete the task associated to this TODO comment. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)

File: codec/src/main/java/io/netty/handler/codec/MessageToMessageEncoder.java [hunk 6]
```diff
@@ -66,68 +64,31 @@ protected MessageToMessageEncoder() {
...
-                        } else {
-                            p = ctx.newPromise();
-                        }
-                        ctx.write(out.get(i), p);
-                    }
-                    ctx.write(out.get(sizeMinusOne), promise);
-                }
-                out.recycle();
+            if (out == null) {
+                return;
```
Comment (netkins) [inline codec/src/main/java/io/netty/handler/codec/MessageToMessageEncoder.java]: BLOCKER Remove this return statement from this finally block. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1143)

File: transport/src/main/java/io/netty/channel/DefaultChannelPromiseAggregatorFactory.java (hunk 1 above)
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/DefaultChannelPromiseAggregatorFactory.java]: @trustin - this looks like a false positive (rhs is not boolean).

**Comment by Scottmitch** [comment]
@normanmaurer @trustin - Ping.

**Comment by normanmaurer** [comment]
@Scottmitch I'm not sure I really like this a lot. It feels more "clumsy" in terms of API compared to the old `MessageToMessageEncoder`. Also did you check that this really gives any improvements ?

**Comment by Scottmitch** [comment]
@normanmaurer - See benchmark results in https://github.com/netty/netty/pull/5214#issuecomment-217034020. The benefits are more pronounced as the size of the redis command grows (~45% improvement demonstrated in the benchmark).

The API allows the encoders to directly write to the channel ... IMO the prior approach to hide writing to the channel from the encoders just because of the promise creation could also be considered "clumsy".  The new API makes it so the parameters in `ChannelOutboundHandler.write` and `TypeSensitiveMessageEncoder.encode` serve roughly the same purpose, and does not require knowledge of what will be done with the `List` lower levels.

File: transport/src/main/java/io/netty/channel/DefaultChannelPromiseAggregatorFactory.java (hunk 1 above)
Comment (trustin) [inline transport/src/main/java/io/netty/channel/DefaultChannelPromiseAggregatorFactory.java]: Yeah, probably a bug in the analyzer.

**Comment by trustin** [comment]
It's great that it improves the performance, but I'm unsure I like the proposed API change like @normanmaurer said.

Instead of `ChannelPromiseAggregatorFactory`, could we define a subset of `List`?

```java
public interface EncoderOutput {
    void add(Object encoded);
}

private final class MessageToMessageEncoderOutput implements EncoderOutput {
    @Override
    public void add(Object encoded) {
        aggregatedPromise.increment();
        ctx.write(encoded, aggregatedPromise);
    }

    void done() { // Invoked by MessageToMessageEncoder
        // aggregated promise set to success only when the write() for the last add() succeeds.
        aggregatedPromise.done(); 
    }
}
```

I'm somewhat afraid that I might be missing something, so please forgive me if this doesn't make any sense. ;-)

**Comment by normanmaurer** [comment]
Let us close this one for now and re-visit for next major version

**Pull request closed without merging**